src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -90,6 +91,52 @@ class Wavefunction
 
     void flush() const
     {
+        //logic to reorder
+        Fusion fg = fused_.get_fusedgates();
+        auto itemsToFuse = fg.get_items();
+        auto ctrlSet = fg.get_ctrl_set();
+        //getting all qubits to move to lower end of the wfn
+        if (itemsToFuse.size() > 0) {
+            std::vector<unsigned> unionOfAllQubitsInUse;
+            std::unordered_set<unsigned> indicesSet; //set is introduced to guard against duplicate insertion and maintianing original order
+            for (int i = 0; i < itemsToFuse.size(); i++) {
+                auto tempIndices = itemsToFuse[i].get_indices();
+                for (unsigned j = 0; j < tempIndices.size(); j++) {
+                    if (indicesSet.count(tempIndices[j]) == 0) {
+                        unionOfAllQubitsInUse.push_back(tempIndices[j]);
+                        indicesSet.insert(tempIndices[j]);
+                    }
+                }
+            }
+            for (auto it = ctrlSet.begin(); it != ctrlSet.end(); ++it) {
+                if (indicesSet.count(*it) == 0) {
+                    unionOfAllQubitsInUse.push_back(*it);
+                    indicesSet.insert(*it);
+                }
+            }
+            //performing reorder
+            std::vector<qubit_t> indexLocs = qubits(unionOfAllQubitsInUse);
+            for (unsigned i = 0; i < indexLocs.size(); i++)
+            {
+                auto currLoc = indexLocs[i];
+                reorder_wavefunction(currLoc, i);
+                indexLocs = qubits(unionOfAllQubitsInUse);
+            }
+            //keeping old and new location in order to set it appropriately
+            std::unordered_map<unsigned, unsigned> old2newDict;
+            for (unsigned i = 0; i < unionOfAllQubitsInUse.size(); i++) {
+                old2newDict[unionOfAllQubitsInUse[i]] = indexLocs[i];
+            }
+
+            for (int i = 0; i < itemsToFuse.size(); i++) {
```

Comment:
As far as I understand this code, it could be thought as:
1. get current fusion state
2. use the state to reorder the wave function and to create mapping dictionary (the fusion state doesn't change during this step)
3. remap the current fusion state using the dictionary

If this understanding is correct, the code could be refactored to be more efficient by
a. use const references to the current Fusion state as we only need to read it and it doesn't update during step2 above
b. do the collection of qubit ids and reorder the wave function (basically, as it is now)
c. add a single `void remap_fusion(const std::unordered_map<unsigned, unsigned>& old2newDict) {...}` method on fused_ to fully update all corresponding structures.